### ISSUE TITLE
SSH/SFTP/SCP proxy

### ISSUE THREAD

**Issue by mkielak** [ronf/asyncssh#492]
Hi,

I'm writing proxy to ssh service which is utilized by PyCharm and other IDEs as remote interpreter. I have some external server where every user has its account. I want to utilize this server in ssh login process - every user should login to ssh service with his username/password same as on my web service.
I created ssh proxy according to tips from [this thread](https://github.com/ronf/asyncssh/issues/468) it's working but it's not perfect in my case.
I coudn't proxy sftp traffic which is used by PyCharm ssh interpreter so I had to write factory to it.
My proxy is ran as another microservice so I have to share same volume between containers, mount it and serve.

```python
await asyncssh.create_server(SSHProxy, '', SSH_PORT,
                             server_host_keys=['ssh_host_key'],
                             process_factory=handle_client,
                             line_editor=False,
                             sftp_factory=MyFTPServer,
                             allow_scp=True)

async def handle_client(process):
    async with asyncssh.connect('real_ssh_service', known_hosts=None) as conn:
        upstream_process = await conn.create_process(env=process.env,
                                                     term_size=process.term_size,
                                                     term_modes=process.term_modes,
                                                     subsystem=process.subsystem,
                                                     term_type='xterm-color', 
                                                     command=process.command,
                                                     stdin=process.stdin,
                                                     stdout=process.stdout,
                                                     stderr=process.stderr)
        await upstream_process.wait_closed()

    process.exit(0)
```

I am able to connect to python interpreter using my proxy but it's not getting installed packages and not acting in same was as with direct connection. I think that I should be able to proxy all traffic including sftp to original ssh to get rid of SFTPServer but what else should I configure?

**Comment by ronf** [comment]
Hello,

Regarding SFTP, that requires an 8-bit clean binary channel. So, you should copy through the term_type just as you do the other variables, to avoid creating a PTY on the remote system for SFTP sessions. However, even that probably won't be good enough, as there's code in AsyncSSH which explicitly looks for requests for a subsystem of 'sftp' and makes sure it starts an SFTP server when it sees that. Right now, there's no way to tell AsyncSSH to pass that traffic through as a plain tunnel.

Regarding connecting to a Python interpreter or accessing packages, there's no sharing of any code between the client & server in this case, so any Python interpreter you asked to be run in something like the process.command would have to be installed on the remote system and available to whatever user you were logging in as. The same would go for any Python packages you wanted to use -- they too would have to be installed on the server. I haven't used this feature in PyCharm, so it's possible it does some copying of packages for you, but I've never looked into that.

Also, you may have to use a full pathname in whatever command you ask to run, depending on what path is set by the remote system's login shell. All of this should be true when connecting directly to the "real" SSH service as well, though.

Another thought: forcing a term_type of "xterm-color' also forces a PTY to be created, even if the original client doesn't ask for one, and that could cause confusion on the client. If you pass through the requested term_type from the client, you can avoid this problem and that might make the tunnel more transparent.

**Comment by ronf** [comment]
See my comment above about SFTP requiring an "8-bit clean binary channel. All your I/O operations on the channel must be using `bytes`', not `str`, meaning `encoding` on the channel needs to bee set to `None`. Normally, this is set in the `_init_sftp_server` method of `SSHServerStreamSession`, which is called from `session_started` when it sees the subsystem requested was `'sftp'`.

Even if you want to tunnel the traffic and not parse it yourself as SFTP, you need to make sure the channel you open has `encoding` set to `None`. This should already be the case if you are opening TCP channels, but it would not be the case if you were opening processes or sessions. You should be able to pass `encoding=None` explicitly to `create_process`, though. Just make sure that all your I/O uses `bytes` along with doing that, not `str`.

**Comment by mkielak** [comment]
Thank you for your help!
I was able to create channel which uses bytes instead of str. I was digging inside the code to somehow subclass channel with my own class and set encoding but couldn't do it. I ended with dirty hacks and replacing code in runtime. That's how it looks:

```python

await asyncssh.create_server(SSHProxy, '', SSH_PORT,
                             server_host_keys=['ssh_host_key'],
                             process_factory=handle_client,
                             line_editor=False,
                             acceptor=acceptor)


def acceptor(self):
    self.create_server_channel = types.MethodType(create_server_channel, self)


def create_server_channel(self, *args, **kwargs):
    channel = asyncssh.connection.SSHServerConnection.create_server_channel(self, *args, **kwargs)
    channel.get_subsystem = types.MethodType(get_subsystem, channel)
    channel.set_encoding(None)
    return channel


def get_subsystem(_):
    return None


async def handle_client(process):
    async with asyncssh.connect('real_ssh_server', known_hosts=None, encoding=None) as conn:
        upstream_process = await conn.create_process(
            env=process.env, encoding=None,
            term_size=process.term_size, term_modes=process.term_modes,
            subsystem=process.subsystem, term_type=process.term_type,
            command=process.command,
            stdin=process.stdin, stdout=process.stdout,
            stderr=process.stderr)
        await upstream_process.wait_closed()

    process.exit(0)
```
Is there any way to make it less ugly? Is it easy to add X11 forwarding here?

**Comment by ronf** [comment]
First, a disclaimer: Anything you're likely to do here is going to involve touching lots of AsyncSSH internals, and this is not something I'll be able to provide compatibility for in future releases. Changes of this sort will be at your own risk, and there's a very large chance anything you do here will break in future releases. If I can understand the use case a bit better, perhaps there might be a way to provide an officially supported way to do this, but at this point I'm not sure I understand what you're attempting well enough to suggest one.

Breaking things down a bit:

I don't think you should have to do any hacking of this sort to get all the I/O to be using `bytes`.  You should just be able to set the `encoding` argument on your top-level `asyncssh.connect()` or `conn.create_process()` call to `None` and AsyncSSH will return `bytes` on all read calls and expect `bytes` on all write calls.

Where you may get into trouble here is around the way AsyncSSH will end up calling run_sftp_server() from within session_started() in SSHServerStreamSession, or perhaps even earlier in subsystem_requested() if you don't specify an sftp_factory. I'm guessing that's why you patched channel's get_subsystem(), to trick session_started() into using the SSHProcess path. However, this will also mean you can't query the SSHProcess to determine if the request you're handling is an SFTP request or not. I'm not sure if that matters in your use case, but doing this properly will probably need some other config option which preserves the ability to see what subsystem was requested.

Thinking about this some more, I think your best bet might actually be to fall back to using the lower-level APIs instead of the stream APIs. You've already started down this path by having your SSHProxy class (of type SSHServer). You can define your own SSHServerSession class and return that from session_requested() in your SSHServer, and that lets you define methods like shell_requested(), exec_requested(), and subsystem_requested(). It also lets you define callbacks for whenever you receive data on one of these sessions, and provides access to a low-level channel object you can use to write any received data to. If you use create_connection() instead of create_process() for the server side connection to the "real" SSH server, you can define an SSHClientSession object which has similar callbacks for when you get data from upstream, which could then write the data back to the SSHServerChannel associated with the inbound client.

Hooking in at this level should avoid all the special handling in the AsyncSSH stream APIs, so you shouldn't need to patch anything. It'll require you to associate the client & server sessions with one another so that each has access to the other's channel object, and you'll need to implement methods like data_received() and eof_received() to pass data back and forth, but in the end that should be much cleaner than what you're trying to do right now.

Regarding X11 forwarding, there is currently no code in AsyncSSH which would let you read and write bytes on an X11 channel yourself. So, I don't think there's be any way to forward those bytes between two different SSH connections with either the low-level or high-level APIs that exist right now. The only thing supported is forwarding local X11 client connections to remote X11 servers via an SSH tunnel and vice-versa.

**Comment by ronf** [comment]
You're still depending on some AsyncSSH internals here (_conn, _session_factory in SSHProcess), so I can't promise these internals won't change at some point. However, I'm glad you were able to get something working here...

**Comment by mkielak** [comment]
Yes I was able to get something working, however I have problem with `connection_requested` callback - it isn't called if I add it to my `ServerSession` class. Do you know why?

**Comment by ronf** [comment]
The `connection_requested` method is meant to be defined on an `SSHServer` object, not an `SSHServerSession`. It is similar to `session_requested` in that regard. In your example, you'd define it on your `SSHProxy` class. This method would be called when an SSH client of your proxy was trying to make an outbound TCP forwarding request, and the result returned by `connection_requested` would decide if that was allowed or not and potentially specify an alternate `TCPSession` or stream handler function for the associated connection. Is that what you're looking for here?

**Comment by mkielak** [comment]
Yes, you're right I defined it in my `SSHProxy` class and when I opened ssh tunnel `server_requested` callback was called but `connection_requested` didn't.

**Comment by ronf** [comment]
If you run OpenSSH as a client with "-L" to set up local port forwarding but not "-N", you'll see a call to `server_requested()`, as OpenSSH opens a shell or exec session in addition to setting up a listener for port forwarding. If you add "-N", you should no longer see the call to `server_requested()`, as that additional session won't be opened.

As for `connection_requested()`, you won't see a call to that when port forwarding is being set up. Instead, you'll see a call each time a new TCP connection is forwarded. So, let's say you did something like "ssh -L 3000:localhost:3000 -N ssh_proxy" to set up a local listener on port 3000. After that, every time you opened a TCP connection to port 3000 on the machine running the SSH client, you should see a call to `connection_requested()` in your proxy, and you can selectively allow or deny those connections individually. There's no call in the proxy when the SSH sets up port forwarding because there's no protocol message from the client at that time. You'll get a call when the client first authenticates the overall SSH connection, but after that you won't see anything after that until it attempts to forward a TCP connection.
